client/lib/catch-js-errors/index.js

```diff
@@ -119,13 +118,18 @@ export default class ErrorLogger {
 		} catch ( e ) {}
 	}
 
-	sendToApi( error ) {
-		const xhr = new XMLHttpRequest();
-		xhr.open( 'POST', 'https://public-api.wordpress.com/rest/v1.1/js-error?http_envelope=1', true );
-		xhr.setRequestHeader( 'Content-type', 'application/x-www-form-urlencoded' );
-		let params =
-			'client_id=39911&client_secret=cOaYKdrkgXz8xY7aysv4fU6wL6sK5J8a6ojReEIAPwggsznj4Cb6mW0nffTxtYT8&error=';
-		params += encodeURIComponent( JSON.stringify( error ) );
-		xhr.send( params );
+	async sendToApi( error ) {
+		const body = new window.FormData();
+		body.append( 'error', JSON.stringify( error ) );
+
+		const response = await window.fetch( 'https://public-api.wordpress.com/rest/v1.1/js-error', {
+			method: 'POST',
+			body,
+		} );
+
+		if ( ! response.ok ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Error: Unable to record the error in Logstash.' );
+		}
```

Comment:
`response.ok` merely checks if the HTTP response status is 2xx. That assumes the request was sent and got a response. But the request can fail at network level, never getting that far. In that case, `await fetch` throws an exception that also needs to be caught.

> I think logging the error in the console is good enough at this point, but let me know if you have other suggestions.

That's fine, even empty `catch {}` would be ok. In the first place, we want to avoid an unhandled rejection/exception.

Reply:
I think that 99% of the requests we're doing in Calypso will fail unhandled if they fail at a network level, but I'm fine to address this if it makes you feel better 😉

Reply:
I've replaced the promise rejection check with a simple try/catch in 7ef03c1785a2899a62fc5ed591f04d9f44f625f6

Reply:
> I think that 99% of the requests we're doing in Calypso will fail unhandled if they fail at a network level

I'd say we're pretty good at catching the errors actually: the data layer requires an `onError` handler to be present: https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/utils.js#L173

Reply:
I'd like to note that at this point in the execution order an unhandled (and likely fatal) error has already been thrown, so it's funny we insist to catch any subsequent errors 😉 

I agree with the general rationale, though, where anything that can throw an error should be caught and handled properly.

Reply:
> I'd say we're pretty good at catching the errors actually: the data layer requires an onError handler to be present: https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/utils.js#L173

If only we were using the data layer for everything 🤔 😉

Reply:
> at this point in the execution order an unhandled (and likely fatal) error has already been thrown, so it's funny we insist to catch any subsequent errors

If the error handler throws an error, it will likely be reported by `tracekit`, too, causing a loop. I didn't test it, but looking at the `tracekit` source code, I don't see any recursion guards. So, I'd say that catching errors here is especially important, not the other way around 🙂 

Also, being aware that rejected promises might be created and catching them properly is a good habit to have, no matter how critical it is in any particular situation.

Thanks for working on this PR. I totally understand you didn't expect to rework such a small patch 6 times before merging 😆

Reply:
I can't say I disagree with that 😉  Being on the "counter" side is good for productive discussions, though, right? 😉 

Thanks again for the thorough reviews! Regardless it was a small change, It's pretty useful to get thorough feedback and discuss the problems first-hand!